### PR TITLE
More robust reflection lookups on non-EF types

### LINKED ISSUE #14878: Check for potentially ambiguous reflection
Reflection on any code that isn't EF should look up methods using name and parameter info to avoid breaks if new overloads are added.

See #14853

### PR DESCRIPTION
**Pull request by roji**
Here are some fixes for reflection lookups to make them more robust - I may not have caught them all. Note that I didn't touch reflection on EF Core's types, or some wonky stuff happening in our tests (e.g. [SqlExceptionFactory.cs](https://github.com/aspnet/EntityFrameworkCore/blob/9afed5819d71ab1751741a591d81aa31d179c60b/test/EFCore.SqlServer.FunctionalTests/TestUtilities/SqlExceptionFactory.cs#L11).

Also simplified some lookups by using `GetRuntimeMethod()` rather than elaborate LINQ checks, it seems robust enough but let me know if there's some issue with that.

Fixes #14878

### CODE REVIEW

File: src/EFCore.Relational/Query/ExpressionTranslators/Internal/StringConcatTranslator.cs [hunk 1]
```diff
@@ -23,12 +23,8 @@ namespace Microsoft.EntityFrameworkCore.Query.ExpressionTranslators.Internal
     /// </summary>
     public class StringConcatTranslator : IExpressionFragmentTranslator
     {
-        private static readonly MethodInfo _stringConcatMethodInfo = typeof(string).GetTypeInfo()
-            .GetDeclaredMethods(nameof(string.Concat))
-            .Single(
-                m => m.GetParameters().Length == 2
-                     && m.GetParameters()[0].ParameterType == typeof(object)
-                     && m.GetParameters()[1].ParameterType == typeof(object));
+        private static readonly MethodInfo _stringConcatMethodInfo
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/ExpressionTranslators/Internal/StringConcatTranslator.cs]: Please revert this change. It has been removed in new pipeline.

Comment (roji) [reply src/EFCore.Relational/Query/ExpressionTranslators/Internal/StringConcatTranslator.cs]: Done, sorry, I'll try to be more careful about touching query code.
